hail/python/hailtop/batch_client/aioclient.py

```diff
@@ -443,46 +452,72 @@ def create_job(self, image, command, env=None, mount_docker_socket=False,
-        group = []
-        group_size = 0
-        for spec in byte_job_specs:
-            n = len(spec)
-            if group_size + n < 8000000 and len(group) < 1000:
-                group.append(spec)
-                group_size += n
-            else:
-                groups.append(group)
-                group = [spec]
-                group_size = n
-        if group:
-            groups.append(group)
+        batch_json = await (await self._client._post('/api/v1alpha/batches/create',
+                                                     json=batch_spec)).json()
+        return Batch(self._client, batch_json['id'], self.attributes)
+
+    MAX_BUNCH_BYTESIZE = 8 * 1024 * 1024 - 512 * 1024  # max request size minus room for headers etc
+    MAX_BUNCH_SIZE = 8 * 1024  # ???
 
-        await bounded_gather(*[functools.partial(self._submit_jobs, batch.id, group)
-                               for group in groups],
-                             parallelism=50)
+    async def submit(self,
+                     max_bunch_bytesize=MAX_BUNCH_BYTESIZE,
+                     max_bunch_size=MAX_BUNCH_SIZE):
+        assert max_bunch_bytesize > 0
+        assert max_bunch_size > 0
+        if self._submitted:
+            raise ValueError("cannot submit an already submitted batch")
 
-        await self._client._patch(f'/api/v1alpha/batches/{batch.id}/close')
-        log.info(f'closed batch {b["id"]}')
+        batch = await self._create()
+        log.info(f'created batch {batch.id}')
+
+        id = batch.id
+
+        if self._byte_job_spec_bunches is None:
+            self._byte_job_spec_bunches = []
+            byte_job_specs = [json.dumps(job_spec).encode('utf-8')
+                              for job_spec in self._job_specs]
+            bunch = []
+            bunch_n_bytes = 0
+            for spec in byte_job_specs:
+                n_bytes = len(spec)
+                assert n_bytes < max_bunch_bytesize, (
+                    f'every job spec must be less than max_bunch_bytesize,'
+                    f' { max_bunch_bytesize }B, but {spec} is larger')
+                if bunch_n_bytes + n_bytes < max_bunch_bytesize and len(bunch) < max_bunch_size:
+                    bunch.append(spec)
+                    bunch_n_bytes += n_bytes
+                else:
+                    self._byte_job_spec_bunches.append(bunch)
+                    bunch = [spec]
+                    bunch_n_bytes = n_bytes
+            if bunch:
+                self._byte_job_spec_bunches.append(bunch)
+
+        n_bunches = len(self._byte_job_spec_bunches)
+        results = await bounded_gather(
+            *[functools.partial(self._submit_jobs, id, bunch_index)
+              for bunch_index in range(n_bunches)],
+            parallelism=50,
+            return_exceptions=1)
+        exceptions = [r for r in results if isinstance(r, BaseException)]
+        if len(exceptions) > 0:
+            assert len(exceptions) == 1
+            raise exceptions[0]
```

Comment:
I think bounded_gather with the defualt return_exceptions=False does the right thing here, no?  Then this filter/check is unneeded.